### PR TITLE
SLM-12 - JWTs should expire at midnight

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/security/JwtServiceTest.kt [hunk 1]
```diff
@@ -39,9 +40,20 @@ class JwtServiceTest {
...
+  @Test
+  fun `the generated JWT should expire at midnight`() {
+    val jwtService = jwtService()
+    val jwt = jwtService.generateToken("some.email@company.com")
+
+    val expiresAt = jwtService.expiresAt(jwt)
+    assertThat(expiresAt).isEqualTo(midnight())
+  }
+
+  private fun midnight() = Instant.now().plus(1, ChronoUnit.DAYS).truncatedTo(ChronoUnit.DAYS)
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/security/JwtServiceTest.kt:52]: The test code looks suspiciously like the production code. Not sure what else to do?

**Comment by nathanrussell-moj-digital** [comment]
I cant remember the detail of the requirement of the story, but as you have it (I think!) the JWT will expire at midnight on the day it is issued. IE. if I get a JWT now (11:35 am), it is valid for just over 12 hours. But if I get a JWT at 23:57 I will get a JWT which is valid for 3 minutes.
Is that what they wanted? I thought there was some chat about midnight at the next Monday, so you potentially had up to 7 days (which sounds too much IMHO). But even then there was some complexity about if you requested the JWT on a Monday it should expire at midnight on the next Monday, so you'd get somewhere between 7 and 8 days?

Or did I get the wrong end of the stick ? (quite probable 😀 )

**Comment by mikehalmamoj** [comment]
> I cant remember the detail of the requirement of the story, but as you have it (I think!) the JWT will expire at midnight on the day it is issued. IE. if I get a JWT now (11:35 am), it is valid for just over 12 hours. But if I get a JWT at 23:57 I will get a JWT which is valid for 3 minutes. Is that what they wanted? I thought there was some chat about midnight at the next Monday, so you potentially had up to 7 days (which sounds too much IMHO). But even then there was some complexity about if you requested the JWT on a Monday it should expire at midnight on the next Monday, so you'd get somewhere between 7 and 8 days?
> 
> Or did I get the wrong end of the stick ? (quite probable 😀 )

The Acceptance Criteria is: `When a user clicks on a magic link, they remain logged in on the same browser for 7 days (rounded up to midnight) and go directly to the ‘Create a list of recipients’ page after entering the service`. We already have a configurable expiry, this PR jsut does the rounding up to midnight.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/security/JwtServiceTest.kt [hunk 2]
```diff
@@ -39,9 +43,21 @@ class JwtServiceTest {
...
+  @Test
+  fun `the generated JWT should expire at midnight on the day of expiry`() {
+    val jwtService = jwtService(
+      expiry = Duration.of(1, ChronoUnit.DAYS),
+      clock = Clock.fixed(Instant.parse("2021-11-26T12:18:05Z"), ZoneId.of("Europe/London"))
+    )
+    val jwt = jwtService.generateToken("some.email@company.com")
+
+    val expiresAt = jwtService.expiresAt(jwt)
+    assertThat(expiresAt).isEqualTo(Instant.parse("2021-11-28T00:00:00Z"))
```
Comment (nathanrussell-moj-digital) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/security/JwtServiceTest.kt:55]: Yeah, much nicer with a fixed clock. Not only is it not duplicating the production code, but its absolutely clear what the expectation is 👍

**Review by nathanrussell-moj-digital — APPROVED** [review]

**Pull request merged**